### PR TITLE
Make CUDA print() atomic

### PR DESCRIPTION
**Pull request by pitrou**
This avoids completely garbled output when printing from multiple threads.
Based on PR #1961.

### CODE REVIEW

File: numba/cuda/printimpl.py [hunk 1]
```diff
@@ -69,18 +53,26 @@ def print_varargs(context, builder, sig, args):
...
-    vprint = nvvmutils.declare_vprint(mod)
-    sep = context.insert_string_const_addrspace(builder, " ")
-    eol = context.insert_string_const_addrspace(builder, "\n")
+    vprint = nvvmutils.declare_vprint(builder.module)
+
+    formats = []
+    values = []
 
     for i, (argtype, argval) in enumerate(zip(sig.args, args)):
+        # XXX The return type is a lie: "print_item" doesn't return
```
Comment (sklam) [inline numba/cuda/printimpl.py]: Will it be better to not use ``@lower("print_item", ...`` and use a custom singledispatch or other ways for dispatching?  Note that "print_item" is originally the implementation for the ``PRINT_ITEM`` bytecode.  I believe the interpreter doesn't generate calls to "print_item" anymore.

Comment (pitrou) [inline numba/cuda/printimpl.py]: I thought about that. The downside is that it makes it harder for people to extend print() for other types. I wonder if there is such a need, though. What do you think?

Comment (sklam) [inline numba/cuda/printimpl.py]: I prefer not violating the ``@lower`` behavior even if it makes it more difficult for people to extend print for CUDA.  Due to the parallel execution model of CUDA, the print will most likely be dealing with scalar types and there are less likely to have custom types.

**Comment by pitrou** [comment]
I've addressed the issue with `@lower` and used `@singledispatch` instead.

**Pull request merged**